deepcopy on dataset with private block fails

Recursion in deep copy when running:
```
    ds = pydicom.data.get_testdata_file("MR_small.dcm", read=True)
    block = ds.private_block(0x000b, "My company 001", create=True)
    ds2 = copy.deepcopy(ds)
```

```
../venv/lib/python3.11/site-packages/pydicom/dataset.py:2803: in _copy_implementation
    copied.__dict__[k] = copy_function(v)
E   RecursionError: maximum recursion depth exceeded while calling a Python object

```

The reason for this recursion may be that class PrivateBlock has a reference to it parent dataset.

**Your environment**
module       | version
------       | -------
platform     | Linux-6.7.9-arch1-1-x86_64-with-glibc2.39
Python       | 3.11.8 (main, Feb 12 2024, 14:50:05) [GCC 13.2.1 20230801]
pydicom      | 2.4.4
gdcm         | _module not found_
jpeg_ls      | _module not found_
numpy        | 1.26.4
PIL          | _module not found_
pylibjpeg    | 2.0.0
openjpeg     | 2.1.1
libjpeg      | 2.0.2

--- a/doc/release_notes/v3.0.0.rst
+++ b/doc/release_notes/v3.0.0.rst
@@ -240,6 +240,8 @@ Fixes
 * Fixed a rare issue with converting pixel data to an :class:`~numpy.ndarray` when
   *Bits Stored* is less than *Bits Allocated* and the unused bits haven't been
   set to an appropriate value for correct interpretation of the data.
+* Fixed a ``RecusionError`` when using :func:`copy.deepcopy` with a dataset containing
+  a private block (:issue:`2025`)
 
 
 Deprecations

--- a/doc/release_notes/v3.0.0.rst
+++ b/doc/release_notes/v3.0.0.rst
@@ -240,7 +240,7 @@ Fixes
 * Fixed a rare issue with converting pixel data to an :class:`~numpy.ndarray` when
   *Bits Stored* is less than *Bits Allocated* and the unused bits haven't been
   set to an appropriate value for correct interpretation of the data.
-* Fixed a ``RecusionError`` when using :func:`copy.deepcopy` with a dataset containing
+* Fixed a ``RecursionError`` when using :func:`copy.deepcopy` with a dataset containing
   a private block (:issue:`2025`)
 
 

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -149,7 +149,10 @@ class PrivateBlock:
         """Return ``True`` if the tag with given `element_offset` is in
         the parent :class:`Dataset`.
         """
-        return self.get_tag(element_offset) in self.dataset
+        if self.dataset:
+            return self.get_tag(element_offset) in self.dataset
+
+        return False
 
     def __getitem__(self, element_offset: int) -> DataElement:
         """Return the data element in the parent dataset for the given element
@@ -172,7 +175,10 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        return self.dataset.__getitem__(self.get_tag(element_offset))
+        if self.dataset:
+            return self.dataset.__getitem__(self.get_tag(element_offset))
+
+        raise RuntimeError("The parent dataset has been deallocated")
 
     def __delitem__(self, element_offset: int) -> None:
         """Delete the tag with the given `element_offset` from the dataset.
@@ -190,7 +196,8 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        del self.dataset[self.get_tag(element_offset)]
+        if self.dataset:
+            del self.dataset[self.get_tag(element_offset)]
 
     def add_new(self, element_offset: int, VR: str, value: object) -> None:
         """Add a private element to the parent :class:`Dataset`.
@@ -210,9 +217,10 @@ class PrivateBlock:
             The value of the data element. See :meth:`Dataset.add_new()`
             for a description.
         """
-        tag = self.get_tag(element_offset)
-        self.dataset.add_new(tag, VR, value)
-        self.dataset[tag].private_creator = self.private_creator
+        if self.dataset:
+            tag = self.get_tag(element_offset)
+            self.dataset.add_new(tag, VR, value)
+            self.dataset[tag].private_creator = self.private_creator
 
     def __getstate__(self) -> dict[str, Any]:
         # When pickling convert from a weak to strong reference

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -42,7 +42,6 @@ from typing import (
     TypeVar,
     overload,
 )
-import weakref
 
 try:
     import numpy

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -218,7 +218,7 @@ class PrivateBlock:
         # When pickling convert from a weak to strong reference
         if self.dataset is not None:
             s = self.__dict__.copy()
-            s['_dsref'] = s['_dsref']()
+            s["_dsref"] = s["_dsref"]()
             return s
 
         return self.__dict__
@@ -226,8 +226,8 @@ class PrivateBlock:
     def __setstate__(self, state: dict[str, Any]) -> None:
         # When unpickling convert from a strong to a weak reference
         self.__dict__.update(state)
-        if self.__dict__['_dsref'] is not None:
-            self.__dict__['_dsref'] = weakref.ref(self.__dict__['_dsref'])
+        if self.__dict__["_dsref"] is not None:
+            self.__dict__["_dsref"] = weakref.ref(self.__dict__["_dsref"])
 
 
 def _dict_equal(a: "Dataset", b: Any, exclude: list[str] | None = None) -> bool:

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -149,10 +149,11 @@ class PrivateBlock:
         """Return ``True`` if the tag with given `element_offset` is in
         the parent :class:`Dataset`.
         """
-        if self.dataset:
-            return self.get_tag(element_offset) in self.dataset
+        if self.dataset is None:
+            raise RuntimeError("The parent dataset has been deallocated")
 
-        return False
+        return self.get_tag(element_offset) in self.dataset
+
 
     def __getitem__(self, element_offset: int) -> DataElement:
         """Return the data element in the parent dataset for the given element
@@ -175,10 +176,10 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        if self.dataset:
-            return self.dataset.__getitem__(self.get_tag(element_offset))
+        if self.dataset is None:
+            raise RuntimeError("The parent dataset has been deallocated")
 
-        raise RuntimeError("The parent dataset has been deallocated")
+        return self.dataset.__getitem__(self.get_tag(element_offset))
 
     def __delitem__(self, element_offset: int) -> None:
         """Delete the tag with the given `element_offset` from the dataset.
@@ -196,8 +197,10 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        if self.dataset:
-            del self.dataset[self.get_tag(element_offset)]
+        if self.dataset is None:
+            raise RuntimeError("The parent dataset has been deallocated")
+
+        del self.dataset[self.get_tag(element_offset)]
 
     def add_new(self, element_offset: int, VR: str, value: object) -> None:
         """Add a private element to the parent :class:`Dataset`.
@@ -217,10 +220,12 @@ class PrivateBlock:
             The value of the data element. See :meth:`Dataset.add_new()`
             for a description.
         """
-        if self.dataset:
-            tag = self.get_tag(element_offset)
-            self.dataset.add_new(tag, VR, value)
-            self.dataset[tag].private_creator = self.private_creator
+        if self.dataset is None:
+            raise RuntimeError("The parent dataset has been deallocated")
+
+        tag = self.get_tag(element_offset)
+        self.dataset.add_new(tag, VR, value)
+        self.dataset[tag].private_creator = self.private_creator
 
     def __getstate__(self) -> dict[str, Any]:
         # When pickling convert from a weak to strong reference

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -115,8 +115,8 @@ class PrivateBlock:
         """
         self.group = key[0]
         self.private_creator = key[1]
-        self.block_start = private_creator_element << 8
         self.dataset = dataset
+        self.block_start = private_creator_element << 8
 
     def get_tag(self, element_offset: int) -> BaseTag:
         """Return the private tag ID for the given `element_offset`.

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -117,11 +117,7 @@ class PrivateBlock:
         self.group = key[0]
         self.private_creator = key[1]
         self.block_start = private_creator_element << 8
-        self._dsref = weakref.ref(dataset)
-
-    @property
-    def dataset(self) -> "Dataset | None":
-        return self._dsref()
+        self.dataset = dataset
 
     def get_tag(self, element_offset: int) -> BaseTag:
         """Return the private tag ID for the given `element_offset`.
@@ -149,9 +145,6 @@ class PrivateBlock:
         """Return ``True`` if the tag with given `element_offset` is in
         the parent :class:`Dataset`.
         """
-        if self.dataset is None:
-            raise RuntimeError("The parent dataset has been deallocated")
-
         return self.get_tag(element_offset) in self.dataset
 
     def __getitem__(self, element_offset: int) -> DataElement:
@@ -175,9 +168,6 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        if self.dataset is None:
-            raise RuntimeError("The parent dataset has been deallocated")
-
         return self.dataset.__getitem__(self.get_tag(element_offset))
 
     def __delitem__(self, element_offset: int) -> None:
@@ -196,9 +186,6 @@ class PrivateBlock:
         KeyError
             If no data element exists at that offset.
         """
-        if self.dataset is None:
-            raise RuntimeError("The parent dataset has been deallocated")
-
         del self.dataset[self.get_tag(element_offset)]
 
     def add_new(self, element_offset: int, VR: str, value: object) -> None:
@@ -219,27 +206,19 @@ class PrivateBlock:
             The value of the data element. See :meth:`Dataset.add_new()`
             for a description.
         """
-        if self.dataset is None:
-            raise RuntimeError("The parent dataset has been deallocated")
-
         tag = self.get_tag(element_offset)
         self.dataset.add_new(tag, VR, value)
         self.dataset[tag].private_creator = self.private_creator
 
-    def __getstate__(self) -> dict[str, Any]:
-        # When pickling convert from a weak to strong reference
-        if self.dataset is not None:
-            s = self.__dict__.copy()
-            s["_dsref"] = s["_dsref"]()
-            return s
+    def __deepcopy__(self, memo: Any) -> "PrivateBlock":
+        copied = self.__class__(
+            (self.group, self.private_creator),
+            self.dataset,
+            self.block_start >> 8,
+        )
+        memo[id(self)] = copied
 
-        return self.__dict__
-
-    def __setstate__(self, state: dict[str, Any]) -> None:
-        # When unpickling convert from a strong to a weak reference
-        self.__dict__.update(state)
-        if self.__dict__["_dsref"] is not None:
-            self.__dict__["_dsref"] = weakref.ref(self.__dict__["_dsref"])
+        return copied
 
 
 def _dict_equal(a: "Dataset", b: Any, exclude: list[str] | None = None) -> bool:

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -154,7 +154,6 @@ class PrivateBlock:
 
         return self.get_tag(element_offset) in self.dataset
 
-
     def __getitem__(self, element_offset: int) -> DataElement:
         """Return the data element in the parent dataset for the given element
         offset.

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -120,7 +120,7 @@ class PrivateBlock:
         self._dsref = weakref.ref(dataset)
 
     @property
-    def dataset(self) -> "Dataset":
+    def dataset(self) -> "Dataset | None":
         return self._dsref()
 
     def get_tag(self, element_offset: int) -> BaseTag:

--- a/src/pydicom/dataset.py
+++ b/src/pydicom/dataset.py
@@ -42,6 +42,7 @@ from typing import (
     TypeVar,
     overload,
 )
+import weakref
 
 try:
     import numpy
@@ -115,8 +116,12 @@ class PrivateBlock:
         """
         self.group = key[0]
         self.private_creator = key[1]
-        self.dataset = dataset
         self.block_start = private_creator_element << 8
+        self._dsref = weakref.ref(dataset)
+
+    @property
+    def dataset(self) -> "Dataset":
+        return self._dsref()
 
     def get_tag(self, element_offset: int) -> BaseTag:
         """Return the private tag ID for the given `element_offset`.
@@ -208,6 +213,21 @@ class PrivateBlock:
         tag = self.get_tag(element_offset)
         self.dataset.add_new(tag, VR, value)
         self.dataset[tag].private_creator = self.private_creator
+
+    def __getstate__(self) -> dict[str, Any]:
+        # When pickling convert from a weak to strong reference
+        if self.dataset is not None:
+            s = self.__dict__.copy()
+            s['_dsref'] = s['_dsref']()
+            return s
+
+        return self.__dict__
+
+    def __setstate__(self, state: dict[str, Any]) -> None:
+        # When unpickling convert from a strong to a weak reference
+        self.__dict__.update(state)
+        if self.__dict__['_dsref'] is not None:
+            self.__dict__['_dsref'] = weakref.ref(self.__dict__['_dsref'])
 
 
 def _dict_equal(a: "Dataset", b: Any, exclude: list[str] | None = None) -> bool:

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1152,31 +1152,11 @@ class TestDataset:
     def test_private_block_pickle(self):
         """Test pickling a private block."""
         ds = Dataset()
-        ds.private_block(0x000B, "Foo", create=True)
+        block = ds.private_block(0x000B, "Foo", create=True)
 
         s = pickle.dumps({"ds": ds})
         ds2 = pickle.loads(s)["ds"]
         assert ds2[0x000B0010].value == "Foo"
-
-    def test_private_block_dealloc(self):
-        """Test a private block when the parent is deallocated."""
-        ds = Dataset()
-        block = ds.private_block(0x000B, "Foo", create=True)
-        block.add_new(0x0005, "CS", "ISO_IR 100")
-        del ds
-
-        msg = "The parent dataset has been deallocated"
-        with pytest.raises(RuntimeError, match=msg):
-            block[0x000B1005]
-
-        with pytest.raises(RuntimeError, match=msg):
-            block.add_new(0x0006, "CS", "ISO_IR 100")
-
-        with pytest.raises(RuntimeError, match=msg):
-            del block[0x0005]
-
-        with pytest.raises(RuntimeError, match=msg):
-            0x0005 in block
 
     def test_private_creator_from_raw_ds(self):
         # regression test for #1078

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1144,12 +1144,6 @@ class TestDataset:
     def test_private_block_deepcopy(self):
         """Test deepcopying a private block."""
         ds = Dataset()
-        ds.add_new(0x00080005, "CS", "ISO_IR 100")
-        ds.add_new(0x00090010, "LO", "Creator 1.0")
-        ds.add_new(0x00091001, "SH", "Version1")
-        ds.add_new(0x00090020, "LO", "Creator 2.0")
-        ds.add_new(0x00092001, "SH", "Version2")
-        ds.add_new(0x00092002, "US", 2)
         ds.private_block(0x000B, "Foo", create=True)
 
         ds2 = copy.deepcopy(ds)
@@ -1158,17 +1152,22 @@ class TestDataset:
     def test_private_block_pickle(self):
         """Test pickling a private block."""
         ds = Dataset()
-        ds.add_new(0x00080005, "CS", "ISO_IR 100")
-        ds.add_new(0x00090010, "LO", "Creator 1.0")
-        ds.add_new(0x00091001, "SH", "Version1")
-        ds.add_new(0x00090020, "LO", "Creator 2.0")
-        ds.add_new(0x00092001, "SH", "Version2")
-        ds.add_new(0x00092002, "US", 2)
         ds.private_block(0x000B, "Foo", create=True)
 
         s = pickle.dumps({"ds": ds})
         ds2 = pickle.loads(s)["ds"]
         assert ds2[0x000B0010].value == "Foo"
+
+    def test_private_block_dealloc(self):
+        """Test a private block when the parent is deallocated."""
+        ds = Dataset()
+        block = ds.private_block(0x000B, "Foo", create=True)
+        block.add_new(0x0005, "CS", "ISO_IR 100")
+        del ds
+
+        msg = "The parent dataset has been deallocated"
+        with pytest.raises(RuntimeError, match=msg):
+            block[0x000B1005]
 
     def test_private_creator_from_raw_ds(self):
         # regression test for #1078

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1141,6 +1141,35 @@ class TestDataset:
         item = ds.get_private_item(0x0009, 0x02, "Creator 2.0")
         assert 2 == item.value
 
+    def test_private_block_deepcopy(self):
+        """Test deepcopying a private block."""
+        ds = Dataset()
+        ds.add_new(0x00080005, "CS", "ISO_IR 100")
+        ds.add_new(0x00090010, "LO", "Creator 1.0")
+        ds.add_new(0x00091001, "SH", "Version1")
+        ds.add_new(0x00090020, "LO", "Creator 2.0")
+        ds.add_new(0x00092001, "SH", "Version2")
+        ds.add_new(0x00092002, "US", 2)
+        ds.private_block(0x000B, "Foo", create=True)
+
+        ds2 = copy.deepcopy(ds)
+        assert ds2[0x000B0010].value == "Foo"
+
+    def test_private_block_pickle(self):
+        """Test pickling a private block."""
+        ds = Dataset()
+        ds.add_new(0x00080005, "CS", "ISO_IR 100")
+        ds.add_new(0x00090010, "LO", "Creator 1.0")
+        ds.add_new(0x00091001, "SH", "Version1")
+        ds.add_new(0x00090020, "LO", "Creator 2.0")
+        ds.add_new(0x00092001, "SH", "Version2")
+        ds.add_new(0x00092002, "US", 2)
+        ds.private_block(0x000B, "Foo", create=True)
+
+        s = pickle.dumps({"ds": ds})
+        ds2 = pickle.loads(s)["ds"]
+        assert ds2[0x000B0010].value == "Foo"
+
     def test_private_creator_from_raw_ds(self):
         # regression test for #1078
         ct_filename = get_testdata_file("CT_small.dcm")

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1169,6 +1169,15 @@ class TestDataset:
         with pytest.raises(RuntimeError, match=msg):
             block[0x000B1005]
 
+        with pytest.raises(RuntimeError, match=msg):
+            block.add_new(0x0006, "CS", "ISO_IR 100")
+
+        with pytest.raises(RuntimeError, match=msg):
+            del block[0x0005]
+
+        with pytest.raises(RuntimeError, match=msg):
+            0x0005 in block
+
     def test_private_creator_from_raw_ds(self):
         # regression test for #1078
         ct_filename = get_testdata_file("CT_small.dcm")

--- a/tests/test_dataset.py
+++ b/tests/test_dataset.py
@@ -1152,7 +1152,7 @@ class TestDataset:
     def test_private_block_pickle(self):
         """Test pickling a private block."""
         ds = Dataset()
-        block = ds.private_block(0x000B, "Foo", create=True)
+        ds.private_block(0x000B, "Foo", create=True)
 
         s = pickle.dumps({"ds": ds})
         ds2 = pickle.loads(s)["ds"]

